File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        s=0
        l=[]
        d={}
        for i in nums:
            s+=i
            l.append(s)
        count=0
        print(l)
        flag=0
        for i in l:
            if i==goal:
                count+=1
                if i - goal in d:
                    count+=d[i-goal]
                if i not in d:
                    d[i]=1
                else:
                    d[i]+=1
            else:
                if i-goal in d:
                    count+=d[i-goal]
                if i not in d:
                    d[i]=1
                else:
                    d[i]+=1
        return count

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        vector<int>prefix;
        int s=0;
        for(int i=0;i<nums.size();i++){
            s+=nums[i];
            prefix.push_back(s);
        }
        int count=0;
        map<int,int>mp;
        for(int i=0;i<prefix.size();i++){
            if(prefix[i]==goal){
                count+=1;
                if(mp[prefix[i]-goal]!=0){
                    count+=mp[prefix[i]-goal];
                }
                if(mp[prefix[i]]==0){
                    mp[prefix[i]]=1;
                }
                else{
                    mp[prefix[i]]+=1;
                }
            }
            else{
                if(mp[prefix[i]-goal]!=0){
                    count+=mp[prefix[i]-goal];
                }
                if(mp[prefix[i]]==0){
                    mp[prefix[i]]=1;
                }
                else{
                    mp[prefix[i]]+=1;
                }
            }
        }
        return count;
    }
};